SwiftGit2/Repository.swift

```diff
@@ -558,4 +558,137 @@ final public class Repository {
+					return self.diff(fromTree: nil, toTree: baseTree)
+				}
+
+				var mergeDiff: Result<OpaquePointer?, NSError> = .success(nil)
+				for parent in commit.parents {
+					let diff = self
+						.commit(with: parent.oid)
+						.flatMap { parentCommit in
+							return self.tree(from: parentCommit)
+						}
+						.flatMap { parentTree in
+							return self.diff(fromTree: parentTree, toTree: baseTree)
+						}
+					mergeDiff = mergeDiff
+						.fanout(diff)
+						.flatMap { (mergeDiff, diff) in
+							guard let mergeDiff = mergeDiff else {
+								return .success(diff)
+							}
+							let mergeResult = git_diff_merge(mergeDiff, diff)
+							guard mergeResult == GIT_OK.rawValue else {
+								return .failure(NSError(gitError: mergeResult, pointOfFailure: "git_diff_merge"))
+							}
+							return .success(mergeDiff)
+						}
+				}
+				return .success(mergeDiff.value!!)
+			}
+			.flatMap { diffResult in
+				return self.processDiffDeltas(diffResult)
+			}
+	}
+
+	private func commit(with oid: OID) -> Result<OpaquePointer, NSError> {
+		var unsafeBaseCommit: OpaquePointer? = nil
+		let unsafeBaseOid = UnsafeMutablePointer<git_oid>.allocate(capacity: 1)
+		git_oid_fromstr(unsafeBaseOid, oid.description)
+		let lookupBaseGitResult = git_commit_lookup(&unsafeBaseCommit, self.pointer, unsafeBaseOid)
+		guard lookupBaseGitResult == GIT_OK.rawValue, let unwrapBaseCommit = unsafeBaseCommit else {
+			return Result.failure(NSError(gitError: lookupBaseGitResult, pointOfFailure: "git_commit_lookup"))
+		}
+		git_commit_free(unsafeBaseCommit)
+
+		return Result.success(unwrapBaseCommit)
+	}
+
+	private func diff(fromTree oldTree: OpaquePointer?,
+	                  toTree newTree: OpaquePointer?) -> Result<OpaquePointer, NSError> {
+		var unsafeDiff: OpaquePointer? = nil
+		let diffResult = git_diff_tree_to_tree(&unsafeDiff, self.pointer, oldTree, newTree, nil)
+		guard diffResult == GIT_OK.rawValue, let unwrapDiffResult = unsafeDiff else {
+			return Result.failure(NSError(gitError: diffResult, pointOfFailure: "git_diff_tree_to_tree"))
+		}
+
+		return Result.success(unwrapDiffResult)
+	}
+
+	private func tree(from commit: OpaquePointer) -> Result<OpaquePointer, NSError> {
+		var unsafeTree: OpaquePointer? = nil
+		let treeResult = git_commit_tree(&unsafeTree, commit)
+		guard treeResult == GIT_OK.rawValue, let unwrapTree = unsafeTree else {
+			return Result.failure(NSError(gitError: treeResult, pointOfFailure: "git_commit_tree"))
+		}
+		git_tree_free(unsafeTree)
+
+		return Result.success(unwrapTree)
+	}
+
+	private func processDiffDeltas(_ diffResult: OpaquePointer) -> Result<[Diff.Delta], NSError> {
+		typealias Delta = Diff.Delta
+		var returnDict = [Delta]()
+
+		let count = git_diff_num_deltas(diffResult)
+
+		for i in 0..<count {
+			let delta = git_diff_get_delta(diffResult, i)
+			let gitDiffDelta = Diff.Delta(_: (delta?.pointee)!)
+
+			returnDict.append(gitDiffDelta)
+
+			git_diff_free(OpaquePointer(delta))
```

Comment:
This is incorrect. `git_diff_free` is called on `git_diff`s; `delta` is a `git_diff_delta`.

Reply:
Good catch. According to the docs `The git_diff_delta pointer points to internal data and you do not have to release it when you are done with it.  It will go away when the * git_diff (or any associated git_patch) goes away.`